Write Redis/PostgreSQL to DynamoDB/S3 script.

DFurnes opened the pull request:
### What's this PR do?

This pull request adds a script to migrate our existing shortlinks (in Redis) and clicks (in PostgreSQL) into DynamoDB/S3. It will update existing links or clicks if run multiple times, so it's safe to re-run before and after we swap production traffic.

### How should this be reviewed?

Give the script a look! Any edge cases I may have missed?

### Any background context you want to provide?

I'm testing this out against QA on [an EC2 instance that Sena very kindly provisioned for me](https://www.pivotaltracker.com/story/show/165507221/comments/214967202). Once we're ready to make a production Bertly 2.0 instance, we can use the same box to run this there:

![Screen Shot 2020-06-04 at 12 50 51 PM](https://user-images.githubusercontent.com/583202/83787621-10f94980-a662-11ea-9322-bfdb2ef25452.png)


### Relevant tickets

References [Pivotal #165507221](https://www.pivotaltracker.com/story/show/165507221).

### Checklist

- [ ] This PR has been added to the relevant Pivotal card.
- [ ] Added appropriate feature/unit tests.

katiecrane commented:
> It will update existing links or clicks if run multiple times, so it's safe to re-run before and after we swap production traffic.

Does this mean that it will update but not duplicate already existing links/clicks?

katiecrane commented on scripts/migrate.js:
```diff
@@ -0,0 +1,60 @@
+  await sql.connect();
+
+  // Iterate over our existing links and migrate them to Bertly 2.0's storage:
+  const stream = redis.scanStream({ match: 'bertly:key:*', count: 100 });
```
Is this `count: 100` what lets you get the `chunk` in the next line?

DFurnes replied:
Yep, the `SCAN` Redis command is basically the same idea as Laravel's `chunk()`. It grabs records in a batch of N, and then has a "next page" pointer that you can follow to request more after that. So this is saying to find `bertly:key:*` records in batches of 100.

It returns a [Node.js Stream](https://nodejs.org/api/stream.html) which we can then "iterate" over with `for await`.

katiecrane replied:
Woohoo thanks okay I'm with you on this bit now!

DFurnes commented on scripts/migrate.js:
```diff
@@ -0,0 +1,60 @@
+  for await (const chunk of stream) {
+    for (const link of chunk) {
+      const key = link.replace('bertly:key:', '');
+      const url = await redis.get(link);
```
One thing I realized I haven't thought about yet is [URL normalization](https://github.com/DoSomething/bertly/blob/7dc66f7a1127ad6721653edd6ee421dee3347006/bertly.py#L74) – we do this to make sure that `https://www.dosomething.org` and `https://www.dosomething.org/` would resolve to the same shortened link (since they're the same URL, just formatted differently).

I'll make a ticket to address this before we launch, but shouldn't block the backfill (since all of the existing Bertly 1.0 shortened URLs have already been normalized by that [`url_normalize`](https://pypi.org/project/url-normalize/) Python package).

katiecrane commented on scripts/migrate.js:
```diff
@@ -0,0 +1,60 @@
+  for await (const chunk of stream) {
+    for (const link of chunk) {
+      const key = link.replace('bertly:key:', '');
+      const url = await redis.get(link);
```
I'm confused here...why do you need to `redis.get(link)` if you already have `link`?

DFurnes replied:
Yeah this is kinda weird and maybe could be named more clearly. The `link` variable is the Redis key found by the scan operation above (so it looks like `bertly:key:na9wqk`). We then have to load the value for that key in a second `GET [key]` Redis operation, so `redis.get('bertly:key:na9wqk')` gives us `https://www.example.com`.

DFurnes replied:
I think I'll rename this from `link` to `redisKey`. Does that read a bit better?

katiecrane replied:
OOOOO yes I like that better! Having "key" in the name makes it clearer to me that `redis.get()` would be getting the value associated with that key.

DFurnes replied:
Done!

DFurnes commented:
> Does this mean that it will update but not duplicate already existing links/clicks?

Correct! If it's, say, re-importing `dosome.click/na9wqk`, it'd just overwrite the record with `{ key: "na9wqk" }` for the link & any click records based on their UUID (from the `click_id` column).

katiecrane commented on scripts/migrate.js:
```diff
@@ -0,0 +1,60 @@
+          user_agent: row.user_agent,
+        };
+
+        filesystem().put(`${clickId}.json`, JSON.stringify(payload));
```
I'm confused how this can be re-run without creating duplicates, but this line makes it seem like we _would_ get duplicates? Since there are no checks if that click already exists?

DFurnes replied:
The `put` method writes a file to that path, so if one already exists it'll just overwrite it.

That said, I refactored this in the extracted `migrate-clicks` script to check if the file exists first, because after splitting these we need to know if this is a "new" click to know whether to increment the `counter` field or not.

katiecrane replied:
OH same file name I see!! Smart!

katiecrane commented on scripts/migrate.js:
```diff
@@ -0,0 +1,60 @@
+
+      // Write link record to DynamoDB:
+      const count = result.rows.length;
+      await Link.update({ key }, { url, count });
```
This says its updating so I think I see for this link record part how it wouldn't get duplicated!

DFurnes commented:
I checked in to see how this script was doing on QA: over the past 24 hours or so, we'd only migrated 74,294 shortlinks. 🐌 That seemed awfully sluggish (especially considering that we have millions of links to shorten).

After dropping a few `console.time` loggers through the script, it was clear that we were spending too much time in PostgreSQL – multiple seconds per link. This was because, despite [what this migration might argue](https://github.com/DoSomething/bertly/blob/7dc66f7a1127ad6721653edd6ee421dee3347006/migrations/versions/42a810a76302_add_shortened_index.py#L19), we were missing an index on `shortened` in production!

I created this index by hand to see if we could speed things up a bit:

```sql
CREATE INDEX shortened_index ON clicks (shortened);
```

And so far so good – after just 30 minutes w/ the index, we've already migrated 72,753 links! 🐎

DFurnes commented:
Okay – made a few more changes after letting this run a few times over the weekend:

🔄 I swapped from [`ioredis`](https://github.com/luin/ioredis) to [`redis`](https://www.npmjs.com/package/redis) because it lets us set our `cursor` for the `SCAN` operation (so we can restart if necessary). For readability, I wrote a [helper generator function](https://github.com/DoSomething/bertly/pull/74/files#diff-5471659d276bea52e1044a1d03b83ce4R17-R31) that handles cursor logic. (Unlike [the library](https://github.com/wspringer/redis-async-gen/blob/2d7881cbf628b016c63bf89c1d3221fe3446044d/src/index.js#L7-L14) that I took inspiration from, this code uses a `for` loop rather than recursion so we don't overflow Node's [call stack](https://2ality.com/2014/04/call-stack-size.html).)

↔️ I split the script into separate `migrate-links` and `migrate-clicks` scripts. This let me simplify each script's logic a bit, but also saves a lot of time – most links have zero clicks (and we have a lot of links), and so it saves us time to bypass those `SELECT` queries for click-less links.

⏱️ I removed unnecessary `await`s from the script (so we just continually fire off requests to DynamoDB and S3, rather than waiting for each request to complete sequentially). This allowed us to increase the script's throughput from 209,500 links/hour to 1,445,556 links/hour.

sheyd commented:
@DFurnes I'd already pre-approved this (you have excellent Code Credit Rating 💳 ), but having looked at the latest changes, consider your PR **re-approved**!

:shipit:

katiecrane commented on scripts/migrate-clicks.js:
```diff
@@ -0,0 +1,80 @@
+      await Link.update(
+        { key: row.shortened },
+        { $ADD: { count: 1 } },
+        { condition: new Condition().filter('url').exists() }
```
Why filter on `url` rather than the `key`?

DFurnes replied:
Yeah I think that'd work too... I know the condition would block saving the record if it doesn't pass, but I'm not sure if it's checking against the "original" document or the document after this "update" is made (in which case `key` would exist).

(The idea here is to exclude links that hadn't yet been imported by the `migrate-links` script, which sets `key` and `url`, and so that's what came to mind for me).
